### PR TITLE
test(meshservice): fix flaky tests on resource grace period

### PR DESCRIPTION
**Pull request by jijiechen**
Fixes flaky test at:
https://github.com/kumahq/kuma/actions/runs/10946133312/job/30392313021#step:6:183

The original asserion can overflow the grace period on resouces, now we strictly read the start time from the resource label to prevent the overflow.

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues 
  - N/A
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS 
  - Confirmed
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) 
  - Manual tested
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? 
  - No
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) 
  - No

> Changelog: skip

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/core/resources/apis/meshservice/generate/generator_test.go [hunk 1]
```diff
@@ -266,20 +266,25 @@ var _ = Describe("MeshService generator", func() {
...
 		}, "2s", "100ms").Should(Succeed())
+
 		labelExistedSince := time.Now()
+		err = labelExistedSince.UnmarshalText([]byte(gracePeriodStartedAt))
+		Expect(err).ToNot(HaveOccurred())
 
 		// Before the grace period it still exists and afterwards it eventually
 		// disappears
 		Consistently(func(g Gomega) {
 			if time.Now().Before(labelExistedSince.Add(gracePeriodInterval)) {
```
Comment (michaelbeaumont) [inline pkg/core/resources/apis/meshservice/generate/generator_test.go:284]: Ok so `labelExistedSince` is going to be _earlier_, which makes it less likely that the `Before` check will succeed and the branch will be taken. Seems to be an improvement.

Should we maybe only call `Expect` _after_ we call `Get` and only if the time is still `Before`? Cause AFAICT we still have the possibility of checking `Before` but then the resource is deleted, and the Consistently fails

File: pkg/core/resources/apis/meshservice/generate/generator_test.go [hunk 2]
```diff
@@ -266,20 +266,25 @@ var _ = Describe("MeshService generator", func() {
...
+		gracePeriodStartedAt := ""
 		// Wait until the MeshService has been marked with grace period start
 		Eventually(func(g Gomega) {
 			g.Expect(resManager.Get(context.Background(), ms, store.GetByKey("backend", model.DefaultMesh))).To(Succeed())
-			g.Expect(ms.GetMeta().GetLabels()).To(HaveKey("kuma.io/deletion-grace-period-started-at"))
+			g.Expect(ms.GetMeta().GetLabels()).To(HaveKey(mesh_proto.DeletionGracePeriodStartedLabel))
+			gracePeriodStartedAt = ms.GetMeta().GetLabels()[mesh_proto.DeletionGracePeriodStartedLabel]
 		}, "2s", "100ms").Should(Succeed())
+
 		labelExistedSince := time.Now()
```
Comment (michaelbeaumont) [inline pkg/core/resources/apis/meshservice/generate/generator_test.go:277]: It's confusing that we have both gracePeriodStartedAt and labelExistedSince

Comment (jijiechen) [reply pkg/core/resources/apis/meshservice/generate/generator_test.go:277]: Let me rename them.

File: pkg/core/resources/apis/meshservice/generate/generator_test.go [hunk 3]
```diff
@@ -266,20 +266,26 @@ var _ = Describe("MeshService generator", func() {
...
 		Eventually(func(g Gomega) {
 			g.Expect(resManager.Get(context.Background(), ms, store.GetByKey("backend", model.DefaultMesh))).To(Succeed())
-			g.Expect(ms.GetMeta().GetLabels()).To(HaveKey("kuma.io/deletion-grace-period-started-at"))
+			g.Expect(ms.GetMeta().GetLabels()).To(HaveKey(mesh_proto.DeletionGracePeriodStartedLabel))
+			labelGracePeriodStartedAt = ms.GetMeta().GetLabels()[mesh_proto.DeletionGracePeriodStartedLabel]
 		}, "2s", "100ms").Should(Succeed())
-		labelExistedSince := time.Now()
+
+		gracePeriodStartedAt := time.Now()
+		err = gracePeriodStartedAt.UnmarshalText([]byte(labelGracePeriodStartedAt))
```
Comment (michaelbeaumont) [inline pkg/core/resources/apis/meshservice/generate/generator_test.go:278]:
```suggestion
		gracePeriodStartedAt, err := time.Parse(time.RFC3339, labelGracePeriodStartedAt)
```

Comment (jijiechen) [reply pkg/core/resources/apis/meshservice/generate/generator_test.go:278]: It's actually invoking `parseStrictRFC3339` inside `UnmarshalText`, and it's the reverse method of the `MarshalText` we are using to generating the label.
See https://github.com/kumahq/kuma/blob/ee63fb3e29ed6e71d7cc730b026836267dc662af/pkg/core/resources/apis/meshservice/generate/generator.go#L230

File: pkg/core/resources/apis/meshservice/generate/generator_test.go [hunk 4]
```diff
@@ -266,20 +266,26 @@ var _ = Describe("MeshService generator", func() {
...
 			g.Expect(resManager.Get(context.Background(), ms, store.GetByKey("backend", model.DefaultMesh))).To(Succeed())
-			g.Expect(ms.GetMeta().GetLabels()).To(HaveKey("kuma.io/deletion-grace-period-started-at"))
+			g.Expect(ms.GetMeta().GetLabels()).To(HaveKey(mesh_proto.DeletionGracePeriodStartedLabel))
+			labelGracePeriodStartedAt = ms.GetMeta().GetLabels()[mesh_proto.DeletionGracePeriodStartedLabel]
 		}, "2s", "100ms").Should(Succeed())
-		labelExistedSince := time.Now()
+
+		gracePeriodStartedAt := time.Time{}
+		err := gracePeriodStartedAt.UnmarshalText([]byte(labelGracePeriodStartedAt))
+		Expect(err).ToNot(HaveOccurred())
```
Comment (michaelbeaumont) [inline pkg/core/resources/apis/meshservice/generate/generator_test.go:279]: Woops sorry

```suggestion
		Expect(gracePeriodStartedAt.UnmarshalText([]byte(labelGracePeriodStartedAt))).To(Succeed())
```

File: pkg/core/resources/apis/meshservice/generate/generator_test.go [hunk 5]
```diff
@@ -266,20 +266,25 @@ var _ = Describe("MeshService generator", func() {
...
 		// disappears
 		Consistently(func(g Gomega) {
-			if time.Now().Before(labelExistedSince.Add(gracePeriodInterval)) {
-				g.Expect(resManager.Get(context.Background(), ms, store.GetByKey("backend", model.DefaultMesh))).To(Succeed())
+			err := resManager.Get(context.Background(), ms, store.GetByKey("backend", model.DefaultMesh))
+			if time.Now().Before(gracePeriodStartedAt.Add(gracePeriodInterval)) {
+				g.Expect(err).To(Succeed())
 			}
-		}, "2s", "100ms").Should(Succeed())
+		}, gracePeriodInterval.String(), "50ms").Should(Succeed())
```
Comment (lahabana) [inline pkg/core/resources/apis/meshservice/generate/generator_test.go:287]:
```suggestion
		}, time.Now().Sub(gracePeriedStartedAt).Sub(time.Millis * 50), "50ms").Should(Succeed())
```

**Review by michaelbeaumont — APPROVED** [review]

**Pull request merged**